syncer/syncer_test.go

```diff
@@ -1031,6 +1035,245 @@ func (s *testSyncerSuite) TestCasuality(c *C) {
+
+		// mock downstream db result
+		mock.ExpectBegin()
+		mock.ExpectExec("CREATE DATABASE").WillReturnResult(sqlmock.NewResult(1, 1))
+		mock.ExpectCommit()
+
+		mock.ExpectBegin()
+		mock.ExpectExec("CREATE TABLE").WillReturnResult(sqlmock.NewResult(1, 1))
+		mock.ExpectCommit()
+
+		mock.ExpectBegin()
+		e := newMysqlErr(1050, "Table exist")
+		mock.ExpectExec("CREATE TABLE").WillReturnError(e)
+		mock.ExpectRollback()
+
+		// mock get table in first handle RowEvent
+		mock.ExpectQuery("SHOW COLUMNS").WillReturnRows(
+			sqlmock.NewRows([]string{"Field", "Type", "Null", "Key", "Default", "Extra"}).AddRow("id", "int", "NO", "PRI", null, "").AddRow("age", "int", "NO", "", null, ""))
+		mock.ExpectQuery("SHOW INDEX").WillReturnRows(
+			sqlmock.NewRows([]string{"Table", "Non_unique", "Key_name", "Seq_in_index", "Column_name",
+				"Collation", "Cardinality", "Sub_part", "Packed", "Null", "Index_type", "Comment", "Index_comment"},
+			).AddRow("st", 0, "PRIMARY", 1, "id", "A", 0, null, null, null, "BTREE", "", ""))
+
+		// run sql on upstream db
+		runSQL(_case.testSQLs)
+		// mock expect sql
+		for i, expectSQL := range _case.expectSQLS {
+			mock.ExpectBegin()
+			if strings.HasPrefix(expectSQL.sql, "ALTER") {
+				mock.ExpectExec(expectSQL.sql).WillReturnResult(sqlmock.NewResult(1, int64(i)+1))
+				mock.ExpectCommit()
+				// mock get table after ddl sql exec
+				mock.ExpectQuery("SHOW COLUMNS").WillReturnRows(
+					sqlmock.NewRows([]string{"Field", "Type", "Null", "Key", "Default", "Extra"}).AddRow("id", "int", "NO", "PRI", null, "").AddRow("age", "int", "NO", "", null, "").AddRow("name", "varchar", "NO", "", null, ""))
+				mock.ExpectQuery("SHOW INDEX").WillReturnRows(
+					sqlmock.NewRows([]string{"Table", "Non_unique", "Key_name", "Seq_in_index", "Column_name",
+						"Collation", "Cardinality", "Sub_part", "Packed", "Null", "Index_type", "Comment", "Index_comment"},
+					).AddRow("st", 0, "PRIMARY", 1, "id", "A", 0, null, null, null, "BTREE", "", ""))
+			} else {
+				// change insert to replace because of safe mode
+				mock.ExpectExec(expectSQL.sql).WithArgs(expectSQL.args...).WillReturnResult(sqlmock.NewResult(1, 1))
+				mock.ExpectCommit()
+			}
+		}
+		ctx, cancel := context.WithCancel(context.Background())
+		resultCh := make(chan pb.ProcessResult)
+
+		go syncer.Process(ctx, resultCh)
+
+		go func() {
+			// sleep to ensure ddlExecInfo.Send() happen after ddlExecInfo.Renew() in Process
+			// because Renew() will generate new channel, Send() may send to old channel due to goroutine schedule
+			time.Sleep(1 * time.Second)
+			// mock permit exec ddl request from dm-master
+			req := &DDLExecItem{&pb.ExecDDLRequest{Exec: true}, make(chan error, 1)}
+			syncer.ddlExecInfo.Send(ctx, req)
+		}()
+
+		select {
+		case r := <-resultCh:
+			c.Assert(len(r.Errors), Equals, 0)
+		case <-time.After(2 * time.Second):
+		}
+		cancel()
+
+		c.Assert(syncer.checkpoint.GlobalPoint().Pos, Equals, _case.globalCheckPoint.Pos)
```

Comment:
@amyangfei After I check the code, I found three ways of flushing the globalPoint
```
1. get flush job
2. get ddl job
3. time.Now() - last globalSavePointTime > 30 * time.Second
```

And in current test case syncer.globalPoint didn't flush after this context canceled when blocking in s.streamer.GetEvent(), so this flushedCheckPoint may not equal to globalPoint. So I only checked globalPoint.Pos.
I'm not sure this is the right way to check the point? Any suggest for check the globalPoint?

Reply:
1. when `syncer` unit pauses or stops, it will dispatch flush job and flush checkpoints, we can use this mechanism.
2. we can add some failpoint injection variable to control the flush period.

Besides we can get the binlog position of last XID from the upstream database and compare it with the global point instead of the hard code position

Reply:
What is the proper way to get last XID from upstream database? I tried the `show master status` , retrived the `		syncer.syncer.GetNextPosition()` which is greater than flushedGlobalPoint

Reply:
It is not easy to get last XID position from upstream directly, but we can query `show binlog events from xxx` from upstream and make sure the global point equals to the last XID position, where `xxx` should be a valid position earlier than `global point`